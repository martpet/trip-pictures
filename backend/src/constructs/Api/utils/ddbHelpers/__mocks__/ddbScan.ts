export const ddbScan = vi
  .fn()
  .mockName('ddbScan')
  .mockResolvedValue('dummyDdbScanResponse');
