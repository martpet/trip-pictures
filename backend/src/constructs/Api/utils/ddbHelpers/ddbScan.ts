import {
  DynamoDBClient,
  DynamoDBDocumentClient,
  ScanCommand,
  TableOptions,
} from 'lambda-layer';

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({}));

type Props<ResultItem> = {
  tableOptions: TableOptions;
  attributes: (keyof ResultItem)[];
};

export async function ddbScan<ResultItem>({
  tableOptions,
  attributes,
}: Props<ResultItem>) {
  let items: ResultItem[] = [];

  const scan = async (ExclusiveStartKey?: Record<string, unknown>) => {
    const command = new ScanCommand({
      TableName: tableOptions.tableName,
      ExclusiveStartKey,
      ProjectionExpression: attributes.join(','),
    });

    const { Items, LastEvaluatedKey } = await ddbDocClient.send(command);

    if (Items) {
      items = items.concat(Items as typeof items);
    }

    if (LastEvaluatedKey) {
      await scan(LastEvaluatedKey);
    }
  };

  await scan();

  return items;
}
