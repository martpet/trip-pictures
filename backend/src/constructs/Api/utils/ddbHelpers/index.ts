export * from './ddbBatchGet';
export * from './ddbBatchWrite';
export * from './ddbScan';
