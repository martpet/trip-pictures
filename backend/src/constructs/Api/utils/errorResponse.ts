import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

import { ApiErrorResponseBody } from '~/types';

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

export const errorResponse = (
  traceId: string,
  options: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  } = options;

  const { envName } = globalLambdaProps;

  const errObj: ApiErrorResponseBody['error'] = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errObj.description = description;
  }

  if (error && (exposeError ?? envName !== 'production')) {
    errObj.error = String(error);
  }

  const body: ApiErrorResponseBody = {
    error: errObj,
  };

  console.error({ traceId, error });

  return {
    statusCode,
    body: JSON.stringify(body),
  };
};
