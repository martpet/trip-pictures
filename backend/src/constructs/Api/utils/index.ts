export * from './cookieName';
export * from './ddbHelpers';
export * from './edgeErrorResponse';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
