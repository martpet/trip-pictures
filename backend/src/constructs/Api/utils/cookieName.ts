import { AllCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllCookies) => name;
