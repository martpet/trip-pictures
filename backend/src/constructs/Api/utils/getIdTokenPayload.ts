import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = async (idToken: string): Promise<IdTokenPayload> => {
  const camelcaseKeys = (await import('camelcase-keys')).default;
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
