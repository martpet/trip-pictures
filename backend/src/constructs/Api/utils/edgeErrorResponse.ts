import { getReasonPhrase } from 'http-status-codes';

import { LambdaEdgeResponse } from '~/constructs/Api/types';
import { errorResponse } from '~/constructs/Api/utils/errorResponse';

export const edgeErrorResponse = (
  ...errorResponseArgs: Parameters<typeof errorResponse>
): LambdaEdgeResponse => {
  const { statusCode, body } = errorResponse(...errorResponseArgs);

  const response: LambdaEdgeResponse = {
    status: statusCode,
    statusDescription: getReasonPhrase(statusCode),
    headers: { 'content-type': [{ value: 'application/json' }] },
    body,
  };

  return response;
};
