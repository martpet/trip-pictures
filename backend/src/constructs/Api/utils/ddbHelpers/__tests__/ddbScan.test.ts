import { DynamoDBDocumentClient, ScanCommand } from '@aws-sdk/lib-dynamodb';
import { TableOptions } from 'aws-cdk-lib/aws-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { ddbScan } from '../ddbScan';

vi.mock('~/constructs/Api/utils/errorResponse');

const ddbMock = mockClient(DynamoDBDocumentClient);

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(ScanCommand).resolves({ Items: [{ dummyItemKey: 'dummyItemValue' }] });
});

const tableOptions = {
  tableName: 'dummyTablename',
} as unknown as TableOptions;

const args = [
  {
    tableOptions,
    attributes: ['dummyAttr'],
  },
] as Parameters<typeof ddbScan>;

describe('ddbScan', () => {
  itSendsAwsCommand(ScanCommand, ddbMock, ddbScan, args);
  itResolves(ddbScan, args);

  describe('when "Items" is missing from "ScanCommand" output', () => {
    beforeEach(() => {
      ddbMock.on(ScanCommand).resolves({});
    });
    itResolves(ddbScan, args);
  });

  describe('when "LastEvaluatedKey" exists in "ScanCommand" output', () => {
    beforeEach(() => {
      ddbMock
        .on(ScanCommand)
        .resolvesOnce({
          Items: [{ dummyItemKey1: 'dummyItemValue1' }],
          LastEvaluatedKey: { lastKeyProp: 'lastKeyValue' },
        })
        .resolvesOnce({
          Items: [{ dummyItemKey2: 'dummyItemValue2' }],
        });
    });

    it('sends "ScanCommand" a second time with correct args', async () => {
      await ddbScan(...args);
      expect(ddbMock.commandCalls(ScanCommand)[1].args[0].input).toMatchSnapshot();
    });

    itResolves(ddbScan, args);
  });
});
