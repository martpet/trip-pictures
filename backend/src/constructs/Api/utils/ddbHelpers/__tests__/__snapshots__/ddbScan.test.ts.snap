// Vitest Snapshot v1

exports[`ddbScan > resolves with a correct value 1`] = `
[
  {
    "dummyItemKey": "dummyItemValue",
  },
]
`;

exports[`ddbScan > sends "ScanCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "ExclusiveStartKey": undefined,
  "ProjectionExpression": "dummyAttr",
  "TableName": "dummyTablename",
}
`;

exports[`ddbScan > when "Items" is missing from "ScanCommand" output > resolves with a correct value 1`] = `[]`;

exports[`ddbScan > when "LastEvaluatedKey" exists in "ScanCommand" output > resolves with a correct value 1`] = `
[
  {
    "dummyItemKey1": "dummyItemValue1",
  },
  {
    "dummyItemKey2": "dummyItemValue2",
  },
]
`;

exports[`ddbScan > when "LastEvaluatedKey" exists in "ScanCommand" output > sends "ScanCommand" a second time with correct args 1`] = `
{
  "ExclusiveStartKey": {
    "lastKeyProp": "lastKeyValue",
  },
  "ProjectionExpression": "dummyAttr",
  "TableName": "dummyTablename",
}
`;
